Search json index

sgued opened the pull request:
Based on [this discussion](https://zola.discourse.group/t/make-zola-ability-to-export-search-index-into-json-instead-of-js/1032/3)

Sanity check:

* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/getzola/zola/pulls) for the same update/change?

## Code changes
(Delete or ignore this section for documentation changes)

* [x] Are you doing the PR on the `next` branch?

If the change is a new feature or adding to/changing an existing one:

* [x] Have you created/updated the relevant documentation page(s)?

Description
=========

This PR adds an `index_format` configuration item.
```
[search]
index_format = "javascript" | "json"
```

When set to Javascript, a JSON file is created instead of a `js` file for the index.

This is:

- More convenient. Lazy-loading Json is easier than a script
- More flexible. The JSON file can be loaded with the `fetch` API and be manipulated more easily than a script that sets a global variable for which you don't control the name. While it's not a use case I encountered, I don't think the current setup makes it easy to load multiple search index at the same time, for example for multiple languages.

The docs are updated to 
- Document the feature
- Use the feature and lazy-load the index, saving more than 1MB (200KB with compression) from loading when just loading the site.

sgued committed:
search: Add support for a JSON index

sgued committed:
docs: Document JSON index for search

sgued committed:
docs: Use lazy-loaded JSON index

Keats commented on components/config/src/config/search.rs:
```diff
@@ -1,5 +1,19 @@
+    #[serde(rename = "json")]
+    Json,
+    #[serde(rename = "javascript")]
+    Javascript,
```
We should likely prefix them with elasticlunr as it would the format for those and we want to avoid breaking changes later if we add more format.

sgued replied:
Done in c716bfcdba9ff69e097b7372b59fe192435a7ae7

Keats commented on components/config/src/config/search.rs:
```diff
@@ -1,5 +1,19 @@
+pub enum IndexFormat {
+    #[serde(rename = "json")]
+    Json,
+    #[serde(rename = "javascript")]
```
you can tell serde to change casing for everything in one attribute: https://serde.rs/container-attrs.html#rename_all

sgued replied:
Done in c716bfcdba9ff69e097b7372b59fe192435a7ae7

sgued committed:
Add elasticlunr prefix to search engine format configuration

This will be useful if support for more search libraries are added in the future
